File: 0345-reverse-vowels-of-a-string/0345-reverse-vowels-of-a-string.js
/**
 * @param {string} s
 * @return {string}
 */
const reverseVowels = function(s) {
    const isVowel = c => /[aeiou]/i.test(c);
    const chars = s.split('');
    const swap = (l, r) => {
        [chars[l], chars[r]] = [chars[r], chars[l]];
    }
    
    let l = 0, r = s.length - 1;
    
    while (l < r) {
        if (isVowel(s[l]) && isVowel(s[r])) {
            swap(l, r);
            l++;
            r--;
        }
        else if (isVowel(s[l])) r--;
        else if (isVowel(s[r])) l++;
        else l++, r--;
    }
    
    return chars.join('');
};

File: easy/0345-reverse-vowels-of-a-string/0345-reverse-vowels-of-a-string.cs
public class Solution {
    public string ReverseVowels(string s) 
    {
        
        bool isVowel(char c) 
        {
            return "AEIOUaeiou".IndexOf(c) > -1;
        }
        
        char[] chars = s.ToCharArray();
        
        void swap(int l, int r) 
        {
            char temp = chars[l];
            chars[l] = chars[r];
            chars[r] = temp;
        }
        
        int l = 0, r = s.Length - 1;
        
        while (l < r) 
        {
            bool vowelL = isVowel(s[l]);
            bool vowelR = isVowel(s[r]);
            
            if (vowelL && vowelR)
            {
                swap(l, r);
                l++;
                r--;
            }
            else if (vowelL) r--;
            else if (vowelR) l++;
            else 
            {
                l++;
                r--;
            }
        }
        
        return new string(chars);
    }
}